File: models/record.model.ts
/**
 * 统一记录模型
 * 支持：碎片记录、日记、周记
 */

export enum RecordType {
  QUICK_NOTE = 'quick_note',    // 碎片记录
  JOURNAL = 'journal',           // 日记
  WEEKLY = 'weekly'              // 周记（暂缓实现）
}

export enum ProcessingMode {
  ONLY_RECORD = 'only_record',   // 仅记录（默认）
  WITH_MOOD = 'with_mood',       // 带情绪标签
  WITH_NVC = 'with_nvc'          // NVC 结构化
}

/**
 * 基础记录接口（所有记录类型的共同字段）
 */
export interface BaseRecord {
  id: string
  type: RecordType
  createdAt: string              // ISO 8601 格式
  updatedAt: string

  // 原始内容（永远保留，可编辑）
  transcription: string          // 转写文本/原文
  audioUrl?: string              // 音频 URL（方案 A 中不存储）

  // 可选字段
  processingMode?: ProcessingMode  // 碎片记录的处理模式
  moods?: string[]                 // 情绪标签 ID 列表
  needs?: string[]                 // 需要标签 ID 列表

  // 用户反馈
  patternFeedback?: 'like' | 'dislike' | 'uncertain'
}

/**
 * 碎片记录（QuickNote）
 * 特点：10-30秒语音，可选结构化
 */
export interface QuickNote extends BaseRecord {
  type: RecordType.QUICK_NOTE
  duration?: number              // 录音时长（秒）

  // NVC 分析结果（可选）
  nvc?: {
    observation: string          // 观察
    feelings: FeelingItem[]      // 感受
    needs: string[]              // 需要 ID 列表
    request?: string             // 请求/行动建议
    userConfirmed: boolean       // 用户是否确认
    userModified: boolean        // 用户是否修改
    generatedAt: string
  }
}

/**
 * 日记（Journal）
 * 特点：长记录，原文优先，AI 旁注辅助
 */
export interface Journal extends BaseRecord {
  type: RecordType.JOURNAL
  title?: string                 // 标题（可选，系统可生成）
  summary?: string               // 摘要（1-2 行）
  date: string                   // YYYY-MM-DD（归属日期）
  referencedFragments?: string[] // 引用的碎片 ID 列表

  // AI 旁注（可选，可被用户否定）
  aiAnnotation?: {
    keyFragments: KeyFragment[]
    emotionPeaks: EmotionPeak[]
    suggestedNeeds: string[]
    microExperiment?: MicroExperiment
    reflectionQuestions?: string[]
    userDismissed: boolean       // 用户是否否定此旁注
    generatedAt: string
  }
}

/**
 * 周记（Weekly） - 暂缓实现
 */
export interface Weekly extends BaseRecord {
  type: RecordType.WEEKLY
  weekRange: string              // "2025-01-13 ~ 2025-01-19"
  referencedRecords?: string[]   // 引用的记录 ID 列表
}

/**
 * 联合类型：所有记录类型
 */
export type Record = QuickNote | Journal | Weekly

/**
 * 感受项（情绪 + 强度）
 */
export interface FeelingItem {
  emotion: string                // 情绪 ID
  intensity: 1 | 2 | 3 | 4 | 5   // 强度
}

/**
 * 关键片段
 */
export interface KeyFragment {
  text: string                   // 原文片段（引用）
  category: 'event' | 'emotion' | 'insight'
}

/**
 * 情绪峰值片段
 */
export interface EmotionPeak {
  text: string                   // 片段原文
  emotion: string                // 情绪 ID
  intensity: number
  localNVC?: {                   // 局部 NVC（可选）
    observation?: string
    feelings?: FeelingItem[]
    needs?: string[]
  }
}

/**
 * 微实验
 */
export interface MicroExperiment {
  trigger: string                // 触发信号
  action30s: string              // 30 秒动作
  needFulfilled: string          // 满足的需要
}

File: lib/data/models/record_model.dart
/// 记录数据模型（Hive）
/// 用于本地存储的 Record 模型
library;

import 'package:hive/hive.dart';
import '../../domain/entities/deep_analysis_result.dart';
import '../../domain/entities/record.dart';
import '../../domain/entities/nvc_analysis.dart';

part 'record_model.g.dart';

@HiveType(typeId: 0)
class RecordModel extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  final String type; // 'quick_note' | 'journal' | 'weekly'

  @HiveField(2)
  final String transcription;

  @HiveField(3)
  final DateTime createdAt;

  @HiveField(4)
  final DateTime updatedAt;

  @HiveField(5)
  final String? audioUrl;

  @HiveField(6)
  final double? duration;

  @HiveField(7)
  final String? processingMode;

  @HiveField(8)
  final List<String>? moods;

  @HiveField(9)
  final List<String>? needs;

  @HiveField(10)
  final Map<String, dynamic>? nvc;

  @HiveField(11)
  final String? title;

  @HiveField(12)
  final String? summary;

  @HiveField(13)
  final String? date;

  @HiveField(14)
  final List<String>? referencedFragments;

  @HiveField(15)
  final String? weekRange;

  @HiveField(16)
  final List<String>? referencedRecords;

  @HiveField(17)
  final String? patternFeedback;

  @HiveField(18)
  final List<Map<String, dynamic>>? deepAnalyses;

  RecordModel({
    required this.id,
    required this.type,
    required this.transcription,
    required this.createdAt,
    required this.updatedAt,
    this.audioUrl,
    this.duration,
    this.processingMode,
    this.moods,
    this.needs,
    this.nvc,
    this.title,
    this.summary,
    this.date,
    this.referencedFragments,
    this.weekRange,
    this.referencedRecords,
    this.patternFeedback,
    this.deepAnalyses,
  });

  /// 从 Domain 实体转换
  factory RecordModel.fromEntity(Record entity) {
    return RecordModel(
      id: entity.id,
      type: _recordTypeToJson(entity.type),
      transcription: entity.transcription,
      createdAt: entity.createdAt,
      updatedAt: entity.updatedAt,
      audioUrl: entity.audioUrl,
      duration: entity.duration,
      processingMode: entity.processingMode == null
          ? null
          : _processingModeToJson(entity.processingMode!),
      moods: entity.moods,
      needs: entity.needs,
      nvc: entity.nvc != null
          ? {
              'observation': entity.nvc!.observation,
              'feelings': entity.nvc!.feelings.map((f) => f.toJson()).toList(),
              'needs': entity.nvc!.needs.map((n) => n.toJson()).toList(),
              'request': entity.nvc!.request,
              'insight': entity.nvc!.insight,
              // NVC 智能体的分诊结果；缺失会让历史记录的推荐徽章回退到关键词路由
              'recommendedMethod': entity.nvc!.recommendedMethod,
              'analyzedAt': entity.nvc!.analyzedAt.toIso8601String(),
            }
          : null,
      title: entity.title,
      summary: entity.summary,
      date: entity.date,
      referencedFragments: entity.referencedFragments,
      weekRange: entity.weekRange,
      referencedRecords: entity.referencedRecords,
      patternFeedback: entity.patternFeedback,
      deepAnalyses: entity.deepAnalyses?.map((item) => item.toJson()).toList(),
    );
  }

  /// 转换为 Domain 实体
  Record toEntity() {
    return Record(
      id: id,
      type: _parseRecordType(type),
      transcription: transcription,
      createdAt: createdAt,
      updatedAt: updatedAt,
      audioUrl: audioUrl,
      duration: duration,
      processingMode:
          processingMode != null ? _parseProcessingMode(processingMode!) : null,
      moods: moods,
      needs: needs,
      nvc: nvc != null ? NVCAnalysis.fromJson(_normalizeJsonMap(nvc!)) : null,
      title: title,
      summary: summary,
      date: date,
      referencedFragments: referencedFragments,
      weekRange: weekRange,
      referencedRecords: referencedRecords,
      patternFeedback: patternFeedback,
      deepAnalyses: deepAnalyses
          ?.map((item) => DeepAnalysisResult.fromJson(_normalizeJsonMap(item)))
          .toList(),
    );
  }

  Map<String, dynamic> _normalizeJsonMap(Map<dynamic, dynamic> source) {
    final normalized = <String, dynamic>{};
    for (final entry in source.entries) {
      final key = entry.key.toString();
      final value = entry.value;
      if (value is Map) {
        normalized[key] = _normalizeJsonMap(Map<dynamic, dynamic>.from(value));
      } else if (value is List) {
        normalized[key] = value.map((item) {
          if (item is Map) {
            return _normalizeJsonMap(Map<dynamic, dynamic>.from(item));
          }
          if (item is List) {
            return item.map((nested) {
              if (nested is Map) {
                return _normalizeJsonMap(Map<dynamic, dynamic>.from(nested));
              }
              return nested;
            }).toList();
          }
          return item;
        }).toList();
      } else {
        normalized[key] = value;
      }
    }
    return normalized;
  }

  /// 解析 RecordType
  RecordType _parseRecordType(String type) {
    switch (type) {
      case 'quick_note':
        return RecordType.quickNote;
      case 'journal':
        return RecordType.journal;
      case 'weekly':
        return RecordType.weekly;
      default:
        return RecordType.quickNote;
    }
  }

  /// 解析 ProcessingMode
  ProcessingMode _parseProcessingMode(String mode) {
    switch (mode) {
      case 'only_record':
        return ProcessingMode.onlyRecord;
      case 'with_mood':
        return ProcessingMode.withMood;
      case 'with_nvc':
        return ProcessingMode.withNVC;
      default:
        return ProcessingMode.onlyRecord;
    }
  }

  static String _recordTypeToJson(RecordType type) {
    switch (type) {
      case RecordType.quickNote:
        return 'quick_note';
      case RecordType.journal:
        return 'journal';
      case RecordType.weekly:
        return 'weekly';
    }
  }

  static String _processingModeToJson(ProcessingMode mode) {
    switch (mode) {
      case ProcessingMode.onlyRecord:
        return 'only_record';
      case ProcessingMode.withMood:
        return 'with_mood';
      case ProcessingMode.withNVC:
        return 'with_nvc';
    }
  }

  /// 复制并修改
  RecordModel copyWith({
    String? id,
    String? type,
    String? transcription,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? audioUrl,
    double? duration,
    String? processingMode,
    List<String>? moods,
    List<String>? needs,
    Map<String, dynamic>? nvc,
    String? title,
    String? summary,
    String? date,
    List<String>? referencedFragments,
    String? weekRange,
    List<String>? referencedRecords,
    String? patternFeedback,
    List<Map<String, dynamic>>? deepAnalyses,
  }) {
    return RecordModel(
      id: id ?? this.id,
      type: type ?? this.type,
      transcription: transcription ?? this.transcription,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      audioUrl: audioUrl ?? this.audioUrl,
      duration: duration ?? this.duration,
      processingMode: processingMode ?? this.processingMode,
      moods: moods ?? this.moods,
      needs: needs ?? this.needs,
      nvc: nvc ?? this.nvc,
      title: title ?? this.title,
      summary: summary ?? this.summary,
      date: date ?? this.date,
      referencedFragments: referencedFragments ?? this.referencedFragments,
      weekRange: weekRange ?? this.weekRange,
      referencedRecords: referencedRecords ?? this.referencedRecords,
      patternFeedback: patternFeedback ?? this.patternFeedback,
      deepAnalyses: deepAnalyses ?? this.deepAnalyses,
    );
  }
}
